charms/kserve-controller/src/charm.py

```diff
@@ -0,0 +1,314 @@
+
+    def _on_kserve_controller_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            self.gen_certs(self.model.name, self.app.name)
+
+            self.controller_container.push("/tmp/k8s-webhook-server/serving-certs/tls.crt", Path("/run/cert.pem").read_text(), make_dirs=True)
+            self.controller_container.push("/tmp/k8s-webhook-server/serving-certs/tls.key", Path("/run/server.key").read_text(), make_dirs=True)
+
+            update_layer(
+                self._controller_container_name,
+                self.controller_container,
+                self._controller_pebble_layer,
+                log,
+            )
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                log.error(str(e.msg))
+            else:
+                log.info(str(e.msg))
+
+        # TODO determine status checking if rbac proxy is also up
+        self.unit.status = ActiveStatus()
+
+    def _on_kube_rbac_proxy_ready(self, event):
+        """Define and start a workload using the Pebble API.
+
+        Learn more about Pebble layers at https://github.com/canonical/pebble
+        """
+        try:
+            update_layer(
+                self._rbac_proxy_container_name,
+                self.rbac_proxy_container,
+                self._rbac_proxy_pebble_layer,
+                log,
+            )
+        except ErrorWithStatus as e:
+            self.model.unit.status = e.status
+            if isinstance(e.status, BlockedStatus):
+                log.error(str(e.msg))
+            else:
+                log.info(str(e.msg))
+
+
+        #TODO determine status checking if controller is also up
+        self.unit.status = ActiveStatus()
+
+    def _on_install(self, event):
+        try:
+            self.unit.status = MaintenanceStatus("Creating k8s resources")
+            self.k8s_resource_handler.apply()
+        except ApiError as e:
+            log.error(e)
+            raise
+        self.model.unit.status = ActiveStatus()
+
+    def _on_remove(self, _):
+        self.unit.status = MaintenanceStatus("Removing k8s resources")
+        k8s_resources_manifests = self.k8s_resource_handler.render_manifests()
+        try:
+            delete_many(
+                    self.k8s_resource_handler.lightkube_client,
+                    k8s_resources_manifests,
+                )
+        except ApiError as e:
+            log.warning(f"Failed to delete resources, with error: {e}")
+            raise e
+        self.unit.status = MaintenanceStatus("K8s resources removed")
+
+    def gen_certs(self, namespace, service_name):
+        if Path("/run/cert.pem").exists():
+            log.info("Found existing cert.pem, not generating new cert.")
+            return
+    
+        Path("/run/ssl.conf").write_text(
```

Comment:
Just a though. In my re-write I checked in `ssl.conf.j2` as under `src/templates/` and used it to render the real `ssl.conf`
Kinda makes it more uniform and removes long strings from code.

Reply:
This suggestion will be included in the refactor. You can see it [here](https://github.com/canonical/kserve-operators/blob/KF-803-kserve-refactor/charms/kserve-controller/src/charm.py#L216)